enhancements/ingress/global-options-enable-hsts.md

```diff
@@ -0,0 +1,457 @@
+mentioned in [RFC 6797 Section 8.1](https://tools.ietf.org/html/rfc6797#section-8.1), if an HTTP
+response is received over insecure transport, clients must ignore any present STS headers.
+
+## Proposal
+
+Each goal is addressed in the subsections below.
+
+### Continue to enable HSTS per-route (retain the existing feature functionality)
+The `Strict-Transport-Security` response header is currently applied per-route via an
+annotation called `haproxy.router.openshift.io/hsts_header`.  This annotation has a
+required directive `max-age`, and optional directives `includeSubDomains` and `preload`,
+with the following definitions:
+- `max-age`: delta time range in seconds during which the host is to be regarded as an HSTS host
+  - If set to 0, it negates the effect and the host is no longer regarded as an HSTS host
+  - `maxAgeSeconds` is a time-to-live value, and if it is not refreshed, the HSTS policy for this host will
+  eventually expire
+- `includeSubDomains`: if present, applies to any subdomains of the host's domain name
+- `preload`: if present, tells the client to include this host in its host preload list so that
+it never needs to do an initial load to get the HSTS header (note that this is not defined in RFC 6797
+and is therefore client implementation-dependent)
+
+HSTS is currently implemented as a pattern in the HAProxy template and applied to `edge` and
+`reencrypt` routes that match the `haproxy.router.openshift.io/hsts_header` annotation:
+```gotemplate
+{{- /* hsts header in response: */}}
+ {{- $hstsOptionalTokenPattern := `(?:includeSubDomains|preload)` }}
+ {{- $hstsPattern := printf `(?:%[1]s[;])*max-age=(?:\d+|"\d+")(?:[;]%[1]s)*`  $hstsOptionalTokenPattern -}}
+...
+{{- if matchValues (print $cfg.TLSTermination) "edge" "reencrypt" }}
+    {{- with $hsts := firstMatch $hstsPattern (index $cfg.Annotations "haproxy.router.openshift.io/hsts_header") }}
+  http-response set-header Strict-Transport-Security {{$hsts}}
+    {{- end }}{{/* hsts header */}}
+  {{- end }}{{/* is "edge" or "reencrypt" */}}
+```
+
+#### Enable HSTS for all routes globally, based on the cluster's ingress domain
+To add the header to all HTTP responses from an `edge` or `reencrypt` route, add a new router environment
+variable called `ROUTER_DEFAULT_HSTS`.
+
+Add a new type `IngressControllerHSTSHeaderPolicy` to the API in
+`openshift/api/blob/master/operator/v1/types_ingress.go`, to capture the configuration of the HSTS Policy.
+```go
+// These are strings instead of enums so that they can be passed in easily from the HAProxy template
+const (
+    AllHSTSHeaderPolicyScope         = "All"
+    LimitedHSTSHeaderPolicyScope     = "Limited"
+)
+
+// HSTSDirectiveType is an optional directive that directs HSTS Policy configuration.
+//+kubebuilder:validation:Enum=preload;includeSubDomains
+type HSTSDirectiveType string
+const(
+    // PreloadHSTSDirective directs the client to include this host in its host preload list so that
```

Comment:
The phrase "this host" doesn't work in this context as this is a global setting that applies to all hosts or to all hosts that match some domain in a list.  It probably isn't helpful to spell that all out here, so maybe just change "this host" to "hosts":
```suggestion
    // PreloadHSTSDirective directs the client to include hosts in its host preload list so that
```